src/crypto_utils.ts

```diff
@@ -1,118 +1,143 @@
-    const encryptedR = cipher.update(r)
+    cipher.start()
+    cipher.update(forge.util.createBuffer(r))
+    cipher.finish()
+
+    // Get the encrypted random value 'r' as a Buffer and ensure it's exactly 16 bytes
+    // the node-forge package produces strange results when passing strings
+    const encryptedR = encodeString(cipher.output.data)
 
     // XOR the encrypted random value 'r' with the ciphertext to obtain the plaintext
-    const plaintext = Buffer.alloc(encryptedR.length)
+    const plaintext = new Uint8Array(BLOCK_SIZE)
+
     for (let i = 0; i < encryptedR.length; i++) {
         plaintext[i] = encryptedR[i] ^ ciphertext[i]
     }
 
     return plaintext
 }
 
-export function generateRSAKeyPair(): crypto.KeyPairSyncResult<Buffer, Buffer> {
+export function generateRSAKeyPair(): { publicKey: Uint8Array; privateKey: Uint8Array } {
     // Generate a new RSA key pair
-    return crypto.generateKeyPairSync("rsa", {
-        modulusLength: 2048,
-        publicKeyEncoding: {
-            type: "spki",
-            format: "der", // Specify 'der' format for binary data
-        },
-        privateKeyEncoding: {
-            type: "pkcs8",
-            format: "der", // Specify 'der' format for binary data
-        },
-    })
-}
+    const rsaKeyPair = forge.pki.rsa.generateKeyPair({ bits: 2048 })
 
-export function decryptRSA(privateKey: Buffer, ciphertext: Buffer): Buffer {
-    // Load the private key in PEM format
-    let privateKeyPEM = privateKey.toString("base64")
-    privateKeyPEM = `-----BEGIN PRIVATE KEY-----\n${privateKeyPEM}\n-----END PRIVATE KEY-----`
-    // Decrypt the ciphertext using RSA-OAEP
-    return crypto.privateDecrypt(
-        {
-            key: privateKeyPEM,
-            padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
-            oaepHash: "sha256",
-        },
-        ciphertext
-    )
+    // Convert keys to DER format
+    const privateKey = forge.asn1.toDer(forge.pki.privateKeyToAsn1(rsaKeyPair.privateKey)).data
+    const publicKey = forge.asn1.toDer(forge.pki.publicKeyToAsn1(rsaKeyPair.publicKey)).data
+
+    return {
+        privateKey: encodeString(privateKey),
+        publicKey: encodeString(publicKey)
+    }
 }
 
+export function decryptRSA(privateKey: Uint8Array, ciphertext: string): string {
+    // Convert privateKey from Uint8Array to PEM format
+    const privateKeyPEM = forge.pki.privateKeyToPem(forge.pki.privateKeyFromAsn1(forge.asn1.fromDer(forge.util.createBuffer(privateKey))));
+
+    // Decrypt using RSA-OAEP
+    const rsaPrivateKey = forge.pki.privateKeyFromPem(privateKeyPEM);
+    
+    const decrypted = rsaPrivateKey.decrypt(forge.util.hexToBytes(ciphertext), 'RSA-OAEP', {
+        md: forge.md.sha256.create()
+    });
+
+    const decryptedBytes = encodeString(decrypted)
+
+    const userKey: Array<string> = []
+
+    for (let i = 0; i < decryptedBytes.length; i++) {
+        userKey.push(decryptedBytes[i].toString(16))
+    }
+
+    return userKey.join("")
```

Comment:
join is better best practice then trying to do toString?

Reply:
the toString method does not return the desired output. It just converts the array into a string as-is.